fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,108 @@
+import React from 'react';
+import PageTitle from './page-title';
+import TodoList from './todo-list';
+import TodoForm from './todo-form';
+
+class App extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      todos: []
+    };
+    this.addTodo = this.addTodo.bind(this);
+    this.toggleCompleted = this.toggleCompleted.bind(this);
+  }
+
+  componentDidMount() {
+    this.getAllTodos();
+  }
+
+  getAllTodos() {
+    /**
+     * Use fetch to send a GET request to `/api/todos`.
+     * Then 😉, once the response JSON is received and parsed,
+     * update state with the received todos.
+     */
+    fetch('/api/todos')
+      .then(response => response.json())
+      .then(data => this.setState({ todos: data }));
+  }
+
+  addTodo(newTodo) {
+    /**
+     * Use fetch to send a POST request to `/api/todos`.
+     * Then 😉, once the response JSON is received and parsed,
+     * add the created todo to the state array.
+     *
+     * TIP: Be sure to SERIALIZE the todo in the body with JSON.stringify()
+     * And specify the "Content-Type" header as "application/json"
+     */
+    fetch('/api/todos', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify(newTodo)
+    })
+      .then(response => response.json())
+      .then(data => data)
+      .catch(error => {
+        console.error('Errz', error);
+      });
+  }
+
+  toggleCompleted(todoId) {
+    /**
+     * Find the index of the matching todo in the state array.
+     * Find its "isCompleted" status.
+     * Make a new Object containing the opposite "isCompleted" status.
+     * Use fetch to send a PATCH request to `/api/todos/${todoId}`
+     * Then 😉, once the response JSON is received and parsed,
+     * replace the old todo in state.
+     *
+     * TIP: Be sure to SERIALIZE the updates in the body with JSON.stringify()
+     * And specify the "Content-Type" header as "application/json"
+     *
+     */
+    var mySpecificItem = this.state.todos[todoId - 1];
+    var stat = mySpecificItem.isCompleted;
+    var oldAndKindaLameGenericItem = this.state.todos;
+    if (stat) {
+      stat = false;
+    } else {
+      stat = true;
+    }
+    mySpecificItem.isCompleted = stat;
```

Comment:
We do not want to modify any arrays or objects in state. Per the pseudo code, create a new object.